Feat android enumerate

CDuPlooy opened the pull request:
Hi I'm opening this a bit earlier to get some feedback.  The idea behind this is to expose `Java.enumerateMethods()` to the Objection agent.  This can be used in the REPL via `android hooking enumerate $pattern`. It supports two flags `--json` and `--quiet`

I'm not sure if I need to do some more work to add this to the API or not? 

Any feedback is welcome :) Thanks @egeldenhuys for pointing out my typos.

leonjza commented:
Nice I think this is going to be super useful. Had thoughts on maybe adding it for iOS as well?

I don't think we'd need to make any changes for the API. Think the caller can decide what they want to do with the data.

vvect commented:
I'd definitely enjoy something similar for iOS :) I'll have a look at that and get back to you soonest.

CDuPlooy commented:
Thanks for the feedback! I had a quick look at how `frida-trace` does this, it uses an `ApiResolver`. The cool thing about that is we can then use patterns that match what `frida-trace` does for both Java and ObjectiveC.  I'll add this soon :) 

I was also thinking of adding another option to `hooking watch`, ie `pattern`. Ideal usage would look something like `android hooking watch pattern '*Security*!*toString*'`. We can then pass the matching methods to the RPC calls responsible for watching methods including any additional flags like `--dump-args` etc. I would personally find this very useful, but maybe it'd be more appropriate as a plugin?

leonjza commented:
Yeah sounds great!

CDuPlooy commented:
> I was also thinking of adding another option to hooking watch, ie pattern. Ideal usage would look something like android hooking watch pattern '*Security*!*toString*'. We can then pass the matching methods to the RPC calls responsible for watching methods including any additional flags like --dump-args etc. I would personally find this very useful, but maybe it'd be more appropriate as a plugin?

Managed to add the iOS portion. I think all that's left is to add the above and make sure everything is okay and up to your standards :)

CDuPlooy commented:
Hi! I've added the above, you can see it in action here:

https://asciinema.org/a/pIRcfJylhdbfnnTdCy8p7dfLD
https://asciinema.org/a/vCVXF6o2qiQRQbJkWK4jxyfv2

leonjza commented on agent/src/android/hooking.ts:
```diff
@@ -63,6 +73,39 @@ export namespace hooking {
+                }
+            }
+        }catch {
+            // Some classes don't have constructors, this is a temporary workaround
```
What error occurs that we need a try/catch block here? A comment would be useful to know.

leonjza commented on agent/src/android/hooking.ts:
```diff
@@ -63,6 +73,39 @@ export namespace hooking {
+        const methodName = (method.toGenericString().split('.').filter(part => part.includes('('))[0].split('(')[0])
+        if (methodsAllowList.length === 0 || (methodsAllowList.length > 0 && methodsAllowList.includes(methodName))) {
+          const overloads = clazz[methodName].overloads
+          result[methodName] = {
```
I'll have to pull the PR to test locally, but does it not make sense to just map the full overloads into this object instead of extracting the args & return only? That way we can bubble up other things to the CLI/API if needed.

leonjza commented on objection/commands/android/hooking.py:
```diff
@@ -356,3 +358,95 @@ def search_methods(args: list) -> None:
+                    '(optional: --quiet)', bold=True)
+        return
+
+    shouldDumpJSON = _should_dump_json(args)
```
`snake_case` variable names in the Python code <3

leonjza commented on objection/commands/android/hooking.py:
```diff
@@ -356,3 +358,95 @@ def search_methods(args: list) -> None:
+                        print(f'\t{method}')
+
+
+    targetFile = shouldDumpJSON
```
This is an odd assignment, and I think an example of where the function return/responsibility is confusing based on the name.

CDuPlooy replied:
I have replaced this with:

```python
    if should_dump_json:
        target_file = _get_flag_value('--json', args)
        if target_file:
            results_json['data'] = results
            with open(target_file, 'w') as fd:
                fd.write(json.dumps(results_json))
                click.secho(f'JSON dumped to {should_dump_json}', bold=True)
```

Where the new `_get_flag` function looks like:

```python
def _get_flag_value(flag:str, args: list) -> Optional[str]:
    target = None
    for i in range(len(args)):
        if args[i] == flag:
            target = i + 1

    if target is None:
        return None
    elif target < len(args):
        return args[target]
    else:
        click.secho(f'Could not find specified value for {flag}', bold=True)
        return None
```

Leaving this as unresolved to see if your happy first :)

leonjza commented on objection/commands/ios/hooking.py:
```diff
@@ -319,3 +322,86 @@ def search_method(args: list) -> None:
+
+def enumerate(args: list) -> None:
+    """
+        Enumerates the current Android application for classes and methods.
```
No it doesn't :P Rest of the function also has Android references.

CDuPlooy replied:
ah damn, to think this happened xD Thank you I'll fix this :)

leonjza commented on objection/commands/ios/hooking.py:
```diff
@@ -319,3 +322,86 @@ def search_method(args: list) -> None:
+    # Print the matching methods
+    for klass in data.keys():
+        methods = data[klass]
+        if not shouldBeQuiet:
```
I think this will be noisy, as the watcher already prints what it does. I'd suggest we remove it.

CDuPlooy replied:
No longer relevant as I've removed the enumerate methods, the printing now happens in the relevant `search` functions.

leonjza reviewed (changes requested):
Thanks for this! I did a quick review, without checking out the PR yet (sorry!), so these are face value, me just reading the code comments. Overall I am excited for the changes!

While reviewing, I had 2 thoughts come up. These commands could probably replace what we have now. Using the matchers is simply much more powerful that the current implementation. We could possibly back port the old behaviour in a way if say no `!`, `*` or `[`/`]` characters (or similar are present).

The next (more cheeky) thought was, what if we made the class/method names tab completable. The file system tab completions use small helper functions which we could possibly retrofit to behave appropriately for methods/classes depending on platform. I know these are not the only commands where those would be useful, but something to think about, maybe for a separate feature/pr.

CDuPlooy commented on agent/src/android/hooking.ts:
```diff
@@ -63,6 +73,39 @@ export namespace hooking {
+      const methods = clazz.class.getDeclaredMethods()
+      const result = {}
+      methods.forEach(method => {
+        // This trims out only the function name and uses that to get the overloads
```
Added a new function:

```ts
  // This function takes in a method such as package.class.perform()
  // and extracts only the method name, ie "perform"
  const genericMethodNameToMethodOnly = (fullMethodName: string): string => {
    // Reduces [package, class, perform()] to "perform()"
    const method = fullMethodName.split('.').filter((part: string) => part.includes('('))[0]
    // Now extract everything before the first '('
    return method.substring(0, method.indexOf('('))
  }
```

Modified the inner for loop to read:

```ts
      methods.forEach(method => {
        // This trims out only the function name and uses that to get the overloads
        const methodName = genericMethodNameToMethodOnly(method.toGenericString())
```

CDuPlooy commented:
Hi sorry things have been hectic and my internet is painfully slow. I fixed some bugs and added some new things (also some things to be cleaned up). Once this is merged it should close #477 and #220.  Kindly note that as  far as I can tell the overload list is also not required when using patterns, as it's possible to use the full method signature in the pattern :) Super excited to work on this some more!

CDuPlooy commented:
Closing so I can split this into multiple PRs since it's a bit confusing.
